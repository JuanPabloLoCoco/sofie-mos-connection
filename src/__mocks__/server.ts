import { EventEmitter } from 'events'

import { Server } from 'net'
import { SocketMock } from './socket'
// import { Writable } from 'stream'

const iconv = require('iconv-lite')
iconv.encodingExists('utf16-be')

// Mock the Server class in 'net':
let setTimeoutOrg = setTimeout

let instances: Array<IServerMock> = []

export class ServerMock extends EventEmitter implements Server {
	// Mock implementation:

	maxConnections: number
	connections: number
	listening: boolean
	listenToPort: number

	private _responses: Array<(data: any) => string | Buffer > = []

	constructor () {
		super()

		// @ts-ignore this is comparable with ISocketMock
		instances.push(this)

		this.listen = jest.fn(this.listen)
		this.close = jest.fn(this.close)
		this.address = jest.fn(this.address)
		this.getConnections = jest.fn(this.getConnections)
		this.ref = jest.fn(this.ref)
		this.unref = jest.fn(this.unref)
		// this.addListener = jest.fn(this.addListener)
		// this.emit = jest.fn(this.emit)
		// this.on = jest.fn(this.on)
		// this.once = jest.fn(this.once)
		// this.prependListener = jest.fn(this.prependListener)
		// this.prependOnceListener = jest.fn(this.prependOnceListener)

	}
	static mockClear () {
		instances.splice(0, 9999)
	}
	static get instances () {
		return instances
	}

	listen (port: any) {

		this.listenToPort = port
		setTimeoutOrg(() => {
			this.listening = true
			this.emit('listening', {})
		},1)
		return this
	}
	close (callback?: Function) {
		if (callback) callback()
		setTimeoutOrg(() => {
			this.emit('close')
		},1)
		return this
	}
	address () { return { port: 0, family: 'string', address: 'string' } }
	getConnections (cb: (error: Error | null, count: number) => void) { cb(null, 0) }
	ref () { return this }
	unref () { return this }
	// addListener (event: string, listener: (...args: any[]) => void) { return this }
	// emit (event: string | symbol, ...args: any[]) { return true }
	// on (event: string, listener: (...args: any[]) => void) { return this }
	// once (event: string, listener: (...args: any[]) => void) { return this }
	// prependListener (event: string, listener: (...args: any[]) => void) { return this }
	// prependOnceListener (event: string, listener: (...args: any[]) => void) { return this }
	// ------------------------------------------------------------------------
	// Mock methods:
	mockNewConnection () {
		// "Someone has connected"
		let socket = new SocketMock()
		socket.on('data', (d) => {
			this.emit('data', d)
		})

		this.emit('connection', socket)

		return socket
	}

	mockSentMessage (data: any, encoding: any) {
		encoding = encoding
		if (this._responses.length) {
			// send reply:

			let cb: any = this._responses.shift()
			let msg

			setTimeoutOrg(() => {

				if (typeof cb === 'string') {
					msg = cb
				} else {
					msg = cb(data)
				}

				this.mockReceiveMessage(msg)
			},1)
		}
	}
	mockReceiveMessage (msg: string | Buffer) {
		this.emit('data', msg)
	}
	mockAddReply (cb: (data: any) => string | Buffer) {
		this._responses.push(cb)
	}
	mockClear () {
		this._responses.splice(0, 9999)
	}
	decode (data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode (str: string) {
		return iconv.encode(str, 'utf16-be')
	}
}

export type IServerMock = jest.Mocked<ServerMock>
