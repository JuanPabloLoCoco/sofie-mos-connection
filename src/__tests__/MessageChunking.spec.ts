import { checkMessageSnapshot, clearMocks,
	decode,
	doBeforeAll,
	encode,
	getMosConnection,
	getMosDevice,
	getXMLReply,
	setupMocks
} from './lib'
import {
	IMOSListMachInfo,
	IMOSObject,
	IMOSROAck,
	IMOSROFullStory,
	IMOSRunningOrder,
	MosConnection,
	MosDevice,
	MosString128} from '..'
import { SocketMock } from '../__mocks__/socket'
import { xmlData, xmlApiData } from '../__mocks__/testData'

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})

describe('message chunking', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection
	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onRequestMachineInfo: jest.Mock<any, any>
	let onRequestMOSObject: jest.Mock<any, any>
	let onRequestAllMOSObjects: jest.Mock<any, any>

	let onRunningOrderStory: jest.Mock<any, any>

	beforeAll(async () => {
		SocketMock.mockClear()
		// ServerMock.mockClear()

		mosConnection = await getMosConnection({
			'0': true,
			'1': true // Must support at least one other profile
		}, false)
		mosDevice = await getMosDevice(mosConnection)


		// Profile 0:
		onRequestMachineInfo = jest.fn(() => {
			return Promise.resolve(xmlApiData.machineInfo)
		})
		mosDevice.onRequestMachineInfo((): Promise<IMOSListMachInfo> => {
			return onRequestMachineInfo()
		})
		// Profile 1:
		onRequestMOSObject = jest.fn(() => {
			return Promise.resolve(xmlApiData.mosObj)
		})
		onRequestAllMOSObjects = jest.fn(() => {
			return Promise.resolve([
				xmlApiData.mosObj,
				xmlApiData.mosObj2
			])
		})
		mosDevice.onRequestMOSObject((objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})
		mosDevice.onRequestAllMOSObjects((): Promise<Array<IMOSObject>> => {
			return onRequestAllMOSObjects()
		})

		let roAckReply = () => {
			let ack: IMOSROAck = {
				ID: new MosString128('runningOrderId'),
				Status: new MosString128('OK'),
				Stories: []
			}
			return Promise.resolve(ack)
		}
		// Profile 3:
		onRunningOrderStory = jest.fn(roAckReply)

		mosDevice.onRunningOrderStory ((story: IMOSROFullStory): Promise<IMOSROAck> => {
			return onRunningOrderStory(story)
		})
		let b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper

		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery

	})
	beforeEach(() => {
		onRunningOrderStory.mockClear()

		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		serverSocketMockQuery.mockClear()

	})
	test('init', async () => {
		expect(mosDevice).toBeTruthy()
		expect(socketMockLower).toBeTruthy()
		expect(socketMockUpper).toBeTruthy()
	})

	function chunkSubstr (str: string, size: number) {
		const numChunks = Math.ceil(str.length / size)
		const chunks = new Array(numChunks)

		/* tslint:disable-next-line */
		for (let i = 0, o = 0; i < numChunks; ++i, o += size) {
		  chunks[i] = str.substr(o, size)
		}

		return chunks
	  }

	test('chunks', async () => {
		// Prepare server response
		let mockReply = jest.fn((data: Buffer) => {
			let str = decode(data)
			let messageID = str.match(/<messageID>([^<]+)<\/messageID>/)![1]
			let repl = getXMLReply(messageID, xmlData.roList)
			let chunks = chunkSubstr(repl, 500)
			expect(chunks).toHaveLength(4)
			return chunks.map(c => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		let returnedObj = await mosDevice.sendRequestRunningOrder(xmlApiData.roList.ID!) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		let msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})

	test('chunk around space', async () => {
		// Prepare server response
		let mockReply = jest.fn((data) => {
			let str = decode(data)
			let messageID = str.match(/<messageID>([^<]+)<\/messageID>/)![1]
			let repl = getXMLReply(messageID, xmlData.roList)

			const splitPoint = repl.indexOf('Test MOS')
			expect(splitPoint).not.toEqual(-1)

			const chunks = [
				repl.substr(0, splitPoint + 4),
				repl.substr(splitPoint + 4)
			]
			return chunks.map(c => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		let returnedObj = await mosDevice.sendRequestRunningOrder (xmlApiData.roList.ID!) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		let msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()

		// This is what we split in half
		const storySlug = returnedObj.Stories[1].Slug as MosString128
		expect(storySlug).toBeTruthy()
		expect(storySlug.toString()).toEqual('Test MOS')
	})

	test('junk data before', async () => {
		// Prepare server response
		let mockReply = jest.fn((data) => {
			let str = decode(data)
			let messageID = str.match(/<messageID>([^<]+)<\/messageID>/)![1]
			let repl = getXMLReply(messageID, xmlData.roList)

			const padded = '         JUNK DATA   ' + repl
			return encode(padded)
		})

		socketMockUpper.mockAddReply(mockReply)
		let returnedObj = await mosDevice.getRunningOrder(xmlApiData.roList.ID!) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		let msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})

	test('junk data packet before', async () => {
		// Prepare server response
		let mockReply = jest.fn((data) => {
			let str = decode(data)
			let messageID = str.match(/<messageID>([^<]+)<\/messageID>/)![1]
			let repl = getXMLReply(messageID, xmlData.roList)

			const chunks = [ '         JUNK DATA   ' , repl]
			return chunks.map(c => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		let returnedObj = await mosDevice.getRunningOrder(xmlApiData.roList.ID!) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		let msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})
})
