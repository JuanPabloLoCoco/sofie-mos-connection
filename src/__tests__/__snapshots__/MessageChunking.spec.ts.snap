// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message chunking chunk around space 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking chunk around space 2`] = `
Object {
  "ID": MosString128 {
    "_str": "96857485",
  },
  "Slug": MosString128 {
    "_str": "5PM RUNDOWN",
  },
  "Stories": Array [
    Object {
      "ID": MosString128 {
        "_str": "5983A501:0049B924:8390EF2B",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 645,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000224",
          },
          "Paths": Array [
            Object {
              "Description": "MPEG2 Video",
              "Target": "\\\\servermediaclip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            Object {
              "Description": "WM9 750Kbps",
              "Target": "http://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            Object {
              "Description": "MOS Object",
              "Target": "http://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": MosString128 {
            "_str": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B10",
      },
      "Slug": MosString128 {
        "_str": "Colstat Murder",
      },
    },
    Object {
      "ID": MosString128 {
        "_str": "3854737F:0003A34D:983A0B28",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B11",
      },
      "Slug": MosString128 {
        "_str": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking chunks 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking chunks 2`] = `
Object {
  "ID": MosString128 {
    "_str": "96857485",
  },
  "Slug": MosString128 {
    "_str": "5PM RUNDOWN",
  },
  "Stories": Array [
    Object {
      "ID": MosString128 {
        "_str": "5983A501:0049B924:8390EF2B",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 645,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000224",
          },
          "Paths": Array [
            Object {
              "Description": "MPEG2 Video",
              "Target": "\\\\servermediaclip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            Object {
              "Description": "WM9 750Kbps",
              "Target": "http://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            Object {
              "Description": "MOS Object",
              "Target": "http://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": MosString128 {
            "_str": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B10",
      },
      "Slug": MosString128 {
        "_str": "Colstat Murder",
      },
    },
    Object {
      "ID": MosString128 {
        "_str": "3854737F:0003A34D:983A0B28",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B11",
      },
      "Slug": MosString128 {
        "_str": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking junk data before 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking junk data before 2`] = `
Object {
  "ID": MosString128 {
    "_str": "96857485",
  },
  "Slug": MosString128 {
    "_str": "5PM RUNDOWN",
  },
  "Stories": Array [
    Object {
      "ID": MosString128 {
        "_str": "5983A501:0049B924:8390EF2B",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 645,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000224",
          },
          "Paths": Array [
            Object {
              "Description": "MPEG2 Video",
              "Target": "\\\\servermediaclip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            Object {
              "Description": "WM9 750Kbps",
              "Target": "http://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            Object {
              "Description": "MOS Object",
              "Target": "http://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": MosString128 {
            "_str": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B10",
      },
      "Slug": MosString128 {
        "_str": "Colstat Murder",
      },
    },
    Object {
      "ID": MosString128 {
        "_str": "3854737F:0003A34D:983A0B28",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B11",
      },
      "Slug": MosString128 {
        "_str": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking junk data packet before 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking junk data packet before 2`] = `
Object {
  "ID": MosString128 {
    "_str": "96857485",
  },
  "Slug": MosString128 {
    "_str": "5PM RUNDOWN",
  },
  "Stories": Array [
    Object {
      "ID": MosString128 {
        "_str": "5983A501:0049B924:8390EF2B",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 645,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000224",
          },
          "Paths": Array [
            Object {
              "Description": "MPEG2 Video",
              "Target": "\\\\servermediaclip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            Object {
              "Description": "WM9 750Kbps",
              "Target": "http://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            Object {
              "Description": "MOS Object",
              "Target": "http://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": MosString128 {
            "_str": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B10",
      },
      "Slug": MosString128 {
        "_str": "Colstat Murder",
      },
    },
    Object {
      "ID": MosString128 {
        "_str": "3854737F:0003A34D:983A0B28",
      },
      "Items": Array [
        Object {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": MosString128 {
            "_str": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": Array [
            Object {
              "MosPayload": Object {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": 2,
                "transitionPoint": 463,
              },
              "MosSchema": "http://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": MosString128 {
            "_str": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": MosString128 {
        "_str": "B11",
      },
      "Slug": MosString128 {
        "_str": "Test MOS",
      },
    },
  ],
}
`;
